rcl/src/rcl/tinydir/tinydir.h

```diff
@@ -0,0 +1,820 @@
+/*
```

Comment:
As to the concern of introducing lots of new code in a delicate part of the system: I don't know of a better option than tinydir out there in terms of being lightweight, mature and thoroughly tested. Writing our own from scratch is possible but I can't see how it would be any better?

Regarding introducing a plugin-based system: I agree with the concept and this is probably the way to go, but I don't think it contradicts this change; we would still need a sensible default lookup mechanism in lieu of a plugin, this is meant to be it.

Reply:
> 
> Indeed, I wouldn't suggest we write our own file system library from scratch either. Yet a new library itself is not always the sole source of issues. It could be the added call sites in integrating with the library render the trouble.
> 
> As the code would reside here, we'd also have to maintain it; at a minimum ensuring it is forever in sync with that of upstream. Yet some upstream projects may wither and die, and the burden of code would be ours to bare. This occurred in ROS1, specifically with the XmlRpc++ library that roscpp relied upon, making my life miserable when attempting to port SROS1 to C++ and having to refactor it to support TLS sockets; perhaps the biggest barrier I encountered for security with ROS1.
> 
> https://sourceforge.net/p/xmlrpcpp/bugs
> https://github.com/ros/ros_comm/tree/melodic-devel/utilities/xmlrpcpp
> 
> Again, to quote the first design principle discussed by Saltzer et al.:
> 
> > "a) Economy of mechanism: Keep the design as simple and small as possible. This well-known principle applies to any aspect of a system, but it deserves emphasis for protection mechanisms for this reason: design and implementation errors that result in unwanted access paths will not be noticed during normal use (since normal use usually does not include attempts to exercise improper access paths). As a result, techniques such as line-byline inspection of software and physical examination of hardware that implements protection mechanisms are necessary. For such techniques to be successful, a small and simple design is essential."
> 

I fully understand your aversion. Aren't you comparing apples to oranges though? XMLRPC is a big 5000+ lines of code library, while tinydir is a small header which wraps POSIX/Windows filesystem calls and nothing else. Seems like something we'd want one day anyway.

But again I agree that having the code copied in the repo is not ideal. If we import it in CMake (e.g. as an ExternalProject), is that something you'd be comfortable with? we'd import a [specific release](https://github.com/cxong/tinydir/releases/tag/1.2.4) and would be able to upgrade to newer versions convieniently.

Reply:
> Aren't you comparing apples to oranges though? XMLRPC is a big 5000+ lines of code library, while tinydir is a small header

Good point.

>  If we import it in CMake (e.g. as an ExternalProject), is that something you'd be comfortable with?

I'd be fine including it as long as it could be made a optional plugin dependency so if the external tinydir project was not found in the workspace and it isn't the default behavior, then folks with customer filesystems could still use rcl without relying on tinydir.